qa/zcash/smoke_tests.py

```diff
@@ -0,0 +1,807 @@
+            print('Received %s' % balance)
+            return balance
+
+        time.sleep(1)
+        ttl -= 1
+        if ttl == 0:
+            # Ask user if they want to keep waiting
+            print()
+            print('Balance: %s Expected: %s' % (balance, expected))
+            ret = input('Do you wish to continue waiting? (Y/n) ')
+            if ret.lower() == 'n':
+                print('Address contained %s at timeout' % balance)
+                return balance
+            else:
+                ttl = 300
+
+def wait_and_check_balance(results, case, zcash, addr, expected):
+    balance = wait_for_balance(zcash, addr, expected)
+    if balance != expected and results is not None and len(case) > 0:
+        results[case] = False
+    return balance
+
+def wait_for_txid_operation(zcash, opid, timeout=300):
+    print('Waiting for async operation %s' % opid)
+    result = None
+    for _ in iter(range(timeout)):
+        results = zcash.z_getoperationresult([opid])
+        if len(results) > 0:
+            result = results[0]
+            break
+        time.sleep(1)
+
+    status = result['status']
+    if status == 'failed':
+        print('Operation failed')
+        print(result['error']['message'])
+        return None
+    elif status == 'success':
+        txid = result['result']['txid']
+        print('txid: %s' % txid)
+        return txid
+
+def async_txid_cmd(results, case, zcash, name, args=[]):
+    opid = run_cmd(results, case, zcash, name, args)
+    # Some async commands return a dictionary containing the opid
+    if isinstance(opid, dict):
+        opid = opid['opid']
+    if opid is None:
+        if results is not None and len(case) > 0:
+            results[case] = False
+        return None
+
+    txid = wait_for_txid_operation(zcash, opid)
+    if txid is None:
+        if results is not None and len(case) > 0:
+            results[case] = False
+    return txid
+
+def z_sendmany(results, case, zcash, from_addr, recipients):
+    return async_txid_cmd(results, case, zcash, 'z_sendmany', [
+        from_addr,
+        [{
+            'address': to_addr,
+            'amount': amount,
+        } for (to_addr, amount) in recipients]
+    ])
+
+def check_z_sendmany(results, case, zcash, from_addr, recipients):
+    txid = z_sendmany(results, case, zcash, from_addr, recipients)
+    if txid is None:
+        return [decimal.Decimal('0')]
+    return [wait_and_check_balance(results, case, zcash, to_addr, amount) for (to_addr, amount) in recipients]
+
+def check_z_sendmany_parallel(results, zcash, runs):
+    # First attempt to create all the transactions
+    txids = [(run, z_sendmany(results, run[0], zcash, run[1], run[2])) for run in runs]
+    # Then wait for balance updates caused by successful transactions
+    return [
+        wait_and_check_balance(results, run[0], zcash, to_addr, amount) if txid is not None else 0
+        for (run, txid) in txids
+        for (to_addr, amount) in run[2]]
```

Comment:
I find this a bit confusing. It would be nice to refactor this not to include the nested list comprehension.